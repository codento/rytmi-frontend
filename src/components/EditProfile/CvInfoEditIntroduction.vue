<template>
  <b-row class="my-2">
    <b-col cols="12">
      <h2>Employee introduction</h2>
    </b-col>
    <b-col
      v-for="lang in [{key: 'fi', label: 'Finnish'}, {key: 'en', label: 'English'}]"
      :key="'edit-introduction-elem-' + lang.key"
      sm="6"
    >
      <small>Introduction for CV main page (in {{ lang.label }})</small>
      <b-textarea
        v-model="editedIntroductions[lang.key]"
        placeholder="Introduction for CV main page"
        type="text"
        rows="6"
        :state="introductionIsValid(editedIntroductions[lang.key])"
      />
      <b-form-invalid-feedback
        :id="'input-introduction-feedback-' + lang.key"
        class="text-left"
      >
        <div v-if="editedIntroductions[lang.key].length === 0">
          Introduction cannot be empty
        </div>
        <div v-else>
          Maximum number of characters reached ({{ editedIntroductions[lang.key].length }}/{{ maxIntroductionLength }})
        </div>
      </b-form-invalid-feedback>
      <div v-if="introductionIsValid(editedIntroductions[lang.key])">
        <p class="text-right text-success">
          {{ maxIntroductionLength - editedIntroductions[lang.key].length }}/{{ maxIntroductionLength }}
        </p>
      </div>
    </b-col>
    <b-col cols="12">
      <b-button
        size="sm"
        class="mb-4"
        variant="success"
        :disabled="!introductionIsValid(editedIntroductions.fi) || !introductionIsValid(editedIntroductions.en)"
        @click.prevent="callUpdateProfileAction()"
      >
        Submit changes
      </b-button>
    </b-col>
  </b-row>
</template>

<script>
import { mapActions } from 'vuex'

export default {
  name: 'CvInfoEditIntroduction',
  props: {
    profile: Object
  },
  data () {
    return {
      maxIntroductionLength: 360,
      editedIntroductions: {
        fi: this.profile.introduction ? this.profile.introduction.fi : '',
        en: this.profile.introduction ? this.profile.introduction.en : ''
      }
    }
  },
  computed: {},
  methods: {
    ...mapActions(['updateProfile']),
    introductionIsValid: function (introduction) {
      return introduction.length > 0 && introduction.length <= 360
    },
    callUpdateProfileAction () {
      const editedProfile = Object.assign({}, this.profile)
      editedProfile.introduction = {}
      editedProfile.introduction.fi = this.editedIntroductions.fi
      editedProfile.introduction.en = this.editedIntroductions.en
      this.updateProfile(editedProfile)
        .then((response) => {
          this.$toasted.global.rytmi_success({
            message: 'Profile CV introduction updated.'
          })
        })
        .catch((err) => {
          this.$toasted.global.rytmi_error({
            message: err
          })
        })
    }
  }
}
</script>
