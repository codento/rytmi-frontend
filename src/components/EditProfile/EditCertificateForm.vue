<template>
  <b-form
    id="edit-certificate-form"
  >
    <h3>Add a new certificate</h3>
    <hr>
    <b-row>
      <b-col>
        <b-form-group invalid-feedback="Required">
          <small for="name-fi">Sertifikaatin nimi (suomeksi) *</small>
          <b-input
            id="name-fi"
            v-model="editedCertificate.fi.name"
            type="text"
            placeholder="Sertifikaatin nimi suomeksi"
            :state="inputStates.nameFi"
          />
        </b-form-group>
      </b-col>
      <b-col>
        <b-form-group invalid-feedback="Required">
          <small for="name-en">Certificate name (in English) *</small>
          <b-input
            id="name-en"
            v-model="editedCertificate.en.name"
            type="text"
            placeholder="Name of the certificate in English"
            :state="inputStates.nameEn"
          />
        </b-form-group>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-form-group invalid-feedback="Required">
          <small for="description-fi">Sertifikaatin kuvaus (suomeksi)</small>
          <b-textarea
            id="description-fi"
            v-model="editedCertificate.fi.description"
            type="text"
            rows="2"
            class="form-control"
            placeholder="LisÃ¤tietoa sertifikaatista, jos tarpeen."
          />
        </b-form-group>
      </b-col>
      <b-col>
        <b-form-group invalid-feedback="Required">
          <small for="description-en">Certificate description (in English)</small>
          <b-textarea
            id="description-en"
            v-model="editedCertificate.en.description"
            type="text"
            rows="2"
            class="form-control"
            placeholder="Additional information about the certificate, if needed."
          />
        </b-form-group>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-form-group invalid-feedback="Year is required and must be after 1970 and can't be in the future.">
          <small for="name-fi">Acquisition year *</small>
          <b-input
            id="year"
            v-model="editedCertificate.year"
            type="number"
            placeholder="Year of acquiring the certificate"
            :state="inputStates.year"
            :min="1970"
            :max="new Date().getFullYear()"
          />
        </b-form-group>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-btn
          id="save"
          class="pull-right"
          variant="success"
          @click="save()"
        >
          Save and close
        </b-btn>
        <b-btn
          id="cancel"
          class="pull-right"
          variant="light"
          @click="cancel()"
        >
          Cancel
        </b-btn>
      </b-col>
    </b-row>
  </b-form>
</template>

<script>
import { cloneDeep } from 'lodash'

export default {
  name: 'EditCertificateForm',
  props: {
    certificate: {
      type: Object,
      required: true
    }
  },
  data () {
    return {
      validated: false,
      editedCertificate: cloneDeep(this.certificate)
    }
  },
  computed: {
    inputStates () {
      return {
        nameFi: this.validated ? this.editedCertificate.fi.name.length > 0 : undefined,
        nameEn: this.validated ? this.editedCertificate.en.name.length > 0 : undefined,
        year: this.validated ? this.editedCertificate.year >= 1970 && this.editedCertificate.year <= new Date().getFullYear() : undefined
      }
    },
    formIsValid () {
      const stateArray = []
      for (let entry of Object.entries(this.inputStates)) {
        stateArray.push(entry[1])
      }
      return stateArray.every(item => item)
    }
  },
  methods: {
    save () {
      this.validated = true
      if (this.formIsValid) {
        this.$emit('submit', this.editedCertificate)
      }
    },
    cancel () {
      this.$emit('cancel')
    }
  }
}
</script>

<style scoped>

</style>
