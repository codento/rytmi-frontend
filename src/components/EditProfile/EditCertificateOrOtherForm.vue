<template>
  <b-form
    id="edit-certificate-or-other-form"
  >
    <h3>Add a new certificate or other validation of expertise</h3>
    <hr>
    <b-row>
      <b-col>
        <b-form-group invalid-feedback="Required">
          <small for="name-fi">Sertifikaatin tai muun nimi (suomeksi) *</small>
          <b-input
            id="name-fi"
            v-model="editedCertificateOrOther.fi.name"
            type="text"
            placeholder="Nimi suomeksi"
            :state="inputStates.nameFi"
          />
        </b-form-group>
      </b-col>
      <b-col>
        <b-form-group invalid-feedback="Required">
          <small for="name-en">Certificate's / Other similar item's name (in English) *</small>
          <b-input
            id="name-en"
            v-model="editedCertificateOrOther.en.name"
            type="text"
            placeholder="Name in English"
            :state="inputStates.nameEn"
          />
        </b-form-group>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-form-group invalid-feedback="Required">
          <small for="description-fi">Kuvaus (suomeksi)</small>
          <b-textarea
            id="description-fi"
            v-model="editedCertificateOrOther.fi.description"
            type="text"
            rows="2"
            class="form-control"
            placeholder="LisÃ¤tietoa, jos tarpeen"
          />
        </b-form-group>
      </b-col>
      <b-col>
        <b-form-group invalid-feedback="Required">
          <small for="description-en">Description (in English)</small>
          <b-textarea
            id="description-en"
            v-model="editedCertificateOrOther.en.description"
            type="text"
            rows="2"
            class="form-control"
            placeholder="Additional information, if needed"
          />
        </b-form-group>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-form-group invalid-feedback="Year is required and must be after 1970 and can't be in the future.">
          <small for="name-fi">Acquisition year *</small>
          <b-input
            id="year"
            v-model="editedCertificateOrOther.year"
            type="number"
            placeholder="Year of acquiring the certificate or similar item"
            :state="inputStates.year"
            :min="1970"
            :max="new Date().getFullYear()"
          />
        </b-form-group>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-btn
          id="save-certificate-or-other"
          class="float-right my-2 mx-2"
          variant="primary"
          @click="save()"
        >
          Save and close
        </b-btn>
        <b-btn
          id="close-certificate-or-other-form"
          class="float-left my-2 mx-2"
          variant="light"
          @click="cancel()"
        >
          Cancel
        </b-btn>
      </b-col>
    </b-row>
  </b-form>
</template>

<script>
import { cloneDeep } from 'lodash'

export default {
  name: 'EditCertificateOrOtherForm',
  props: {
    certificateOrOther: {
      type: Object,
      required: true
    }
  },
  data () {
    return {
      validated: false,
      editedCertificateOrOther: cloneDeep(this.certificateOrOther)
    }
  },
  computed: {
    inputStates () {
      return {
        nameFi: this.validated ? this.editedCertificateOrOther.fi.name.length > 0 : undefined,
        nameEn: this.validated ? this.editedCertificateOrOther.en.name.length > 0 : undefined,
        year: this.validated ? this.editedCertificateOrOther.year >= 1970 && this.editedCertificateOrOther.year <= new Date().getFullYear() : undefined
      }
    },
    formIsValid () {
      const stateArray = []
      for (let entry of Object.entries(this.inputStates)) {
        stateArray.push(entry[1])
      }
      return stateArray.every(item => item)
    }
  },
  methods: {
    save () {
      this.validated = true
      if (this.formIsValid) {
        this.$emit('submit', this.editedCertificateOrOther)
      }
    },
    cancel () {
      this.$emit('cancel')
    }
  }
}
</script>

<style scoped>

</style>
