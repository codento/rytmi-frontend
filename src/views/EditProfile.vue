<template>
  <div class="animated fadeIn profile-editor col-sm-12 col-md-10">
    <h1>Edit Profile</h1>
    <hr />
    <h1>Basic Details</h1>
    <hr />
    <template v-if="profile">
      <ProfileForm :profile="profile" />
      <EditLinks :profile="profile" />
      <edit-skills :profileId="profile.id" />
      <edit-projects :profileId="profile.id" />
    </template>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'
import {
  EditLinks,
  EditSkills,
  EditProjects,
  ProfileForm
} from '../components/EditProfile'
import store from '../store'

export default {
  name: 'EditProfile',
  props: [
    'profileId'
  ],
  components: {
    ProfileForm,
    EditLinks,
    EditSkills,
    EditProjects
  },
  computed: {
    ...mapGetters([
      'profileById'
    ]),
    profile () {
      return this.profileById(this.profileId)
    }
  },
  mounted () {
    document.title = 'Rytmi - Edit Profile'
  },
  beforeRouteEnter (to, from, next) {
    store.dispatch('fetchProfileProjects', to.params.profileId)
    next()
  }
}
</script>

<style scoped>
.profile-editor {
  margin: 0 auto;
}
</style>
