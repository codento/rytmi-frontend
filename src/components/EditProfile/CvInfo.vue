<template>
  <div class="my-4 animated fadeIn">
    <h1>Information for CV</h1>
    <hr>
    <CvInfoEditIntroduction :profile="profile" />
    <hr>
    <CvInfoEditEducation :profile="profile" />
  </div>
</template>

<script>
import CvInfoEditIntroduction from './CvInfoEditIntroduction'
import CvInfoEditEducation from './CvInfoEditEducation'

export default {
  name: 'CvInfo',
  components: {
    CvInfoEditIntroduction,
    CvInfoEditEducation
  },
  props: {
    profile: Object
  },
  data () {
    return {}
  },
  computed: {},
  methods: {}
}
</script>
