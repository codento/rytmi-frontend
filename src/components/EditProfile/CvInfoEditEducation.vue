<template>
  <div>
    <b-row class="my-2">
      <b-col cols="12">
        <h2>Education</h2>
      </b-col>
      <b-col class="col-12">
        <b-table
          id="education-table"
          :items="tableItems"
          :fields="fields"
          stacked="sm"
        >
          <template
            slot="school"
            slot-scope="data"
          >
            <span class="table-cell-text text-truncate"> {{ data.value }} </span>
          </template>

          <template
            slot="startYear"
            slot-scope="data"
          >
            {{ data.value }}
          </template>

          <template
            slot="endYear"
            slot-scope="data"
          >
            {{ data.value }}
          </template>

          <template
            slot="degree"
            slot-scope="data"
          >
            <span class="table-cell-text text-truncate"> {{ data.value }} </span>
          </template>

          <template
            slot="major"
            slot-scope="data"
          >
            <span class="table-cell-text text-truncate"> {{ data.value }} </span>
          </template>

          <template
            slot="minor"
            slot-scope="data"
          >
            <span class="table-cell-text text-truncate"> {{ data.value }} </span>
          </template>

          <template
            slot="actions"
            slot-scope="data"
          >
            <b-button-group>
              <b-btn
                :id="'edit-education-item-btn-' + data.index"
                size="sm"
                class="mr-1 table-button"
                variant="primary"
                @click="openEditModal(data)"
              >
                Edit
              </b-btn>
              <b-btn
                :id="'remove-education-item-btn-' + data.index"
                name="remove"
                size="sm"
                class="mr-1 table-button"
                variant="danger"
                @click.stop="removeEducation(data)"
              >
                Remove
              </b-btn>
            </b-button-group>
          </template>
        </b-table>
      </b-col>
      <b-col cols="12">
        <b-button
          id="add-education-btn"
          size="sm"
          class="mr-1"
          variant="primary"
          @click="openEditModal()"
        >
          Add new education
        </b-button>
      </b-col>
    </b-row>
    <b-modal
      ref="edit-education-modal"
      title=""
      size="lg"
      hide-footer
      hide-header
    >
      <CvInfoEditEducationForm
        :initial-values="editedEducationItem"
        @cancel="closeEditModal()"
        @submit="submitEducationForm"
      />
    </b-modal>
  </div>
</template>

<script>
import { cloneDeep } from 'lodash'
import { mapActions, mapGetters } from 'vuex'
import CvInfoEditEducationForm from './CvInfoEditEducationForm'

const educationTemplate = {
  fi: {
    school: '',
    degree: '',
    major: '',
    minor: ''
  },
  en: {
    school: '',
    degree: '',
    major: '',
    minor: ''
  },
  startYear: null,
  endYear: null
}

export default {
  name: 'CvInfoEditEducation',
  components: {
    CvInfoEditEducationForm
  },
  props: {
    profile: Object
  },
  data () {
    return {
      fields: [
        { key: 'school', label: 'Place of education' },
        { key: 'startYear', label: 'Starting year' },
        { key: 'endYear', label: 'Finishing year' },
        { key: 'degree', label: 'Degree' },
        { key: 'major', label: 'Major' },
        { key: 'minor', label: 'Minor' },
        'actions'
      ],
      editedEducation: this.profile.education ? this.profile.education : [],
      editedEducationItem: cloneDeep(educationTemplate),
      editedItemIndex: -1
    }
  },
  computed: {
    ...mapGetters(['currentLanguage']),
    tableItems () {
      return this.editedEducation.map(item => {
        return {
          school: item[this.currentLanguage].school,
          degree: item[this.currentLanguage].degree,
          startYear: item.startYear,
          endYear: item.endYear,
          major: item[this.currentLanguage].major,
          minor: item[this.currentLanguage].minor
        }
      })
    }
  },
  methods: {
    ...mapActions(['updateProfile']),
    removeEducation (item) {
      const confirmation = confirm('Are you sure?')
      if (confirmation) {
        this.editedEducation.splice(item.index, 1)
        this.callUpdateProfileAction()
      }
    },
    openEditModal (item = null) {
      if (item) {
        this.editedEducationItem = cloneDeep(this.editedEducation[item.index])
        this.editedItemIndex = item.index
      } else {
        this.editedEducationItem = cloneDeep(educationTemplate)
        this.editedItemIndex = -1
      }
      this.$refs['edit-education-modal'].show()
    },
    closeEditModal () {
      this.$refs['edit-education-modal'].hide()
    },
    submitEducationForm (submittedData) {
      if (this.editedItemIndex >= 0) {
        this.editedEducation[this.editedItemIndex] = cloneDeep(submittedData)
      } else {
        this.editedEducation.push(submittedData)
      }
      this.callUpdateProfileAction()
    },
    async callUpdateProfileAction () {
      const editedProfile = cloneDeep(this.profile)
      this.editedEducation.sort((a, b) => b.endYear - a.endYear)
      editedProfile.education = this.editedEducation
      try {
        await this.updateProfile(editedProfile)
        this.$toasted.global.rytmi_success({
          message: 'Profile CV education updated.'
        })
        this.$refs['edit-education-modal'].hide()
      } catch (error) {
        this.$toasted.global.rytmi_error({
          message: error
        })
      }
    }
  }
}
</script>

<style scoped >
button.table-button {
  width: 100%;
}
.table-cell-text {
  width: 0;
  min-width: 100%;
  display: inline-block;
  max-width: 150px;
}
.modal-btn {
  margin-top: 0.5rem;
}

</style>
