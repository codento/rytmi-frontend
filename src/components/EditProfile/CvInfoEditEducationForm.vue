<template>
  <div>
    <h3>Edit education data (step {{ currentStep + 1 }} of {{ steps.length }})</h3>
    <hr>
    <b-row class="my-3">
      <b-col
        v-for="step in Array(steps.length).keys()"
        :key="step"
      >
        <div
          :class="step === currentStep ? 'step-title current-step' : 'step-title text-muted'"
        >
          <i
            v-if="step < currentStep"
            class="fa fa-check"
          />
          <span v-else>
            {{ step + 1 }}
          </span>
          <span> {{ steps[step].stepTitle }} </span>
        </div>
      </b-col>
    </b-row>
    <hr>
    <div v-if="steps[currentStep].useFormTemplate">
      <b-row
        v-for="item in formData.filter(item => item.key !== 'startYear' && item.key !== 'endYear')"
        :key="kebabCase(item.key)"
      >
        <b-col cols="12">
          <small :for="'input-'+ kebabCase(item.key) + '-' + currentLanguage">{{ item.label }}</small>
          <b-form-group
            :id="'edit-'+ kebabCase(item.key) + '-' + currentLanguage"
            :description="item.descriptions[currentLanguage]"
            :invalid-feedback="invalidFeedback(item.key, currentLanguage)"
          >
            <b-form-input
              :id="'input-'+ kebabCase(item.key) + '-' + currentLanguage"
              v-model="editedValues[currentLanguage][item.key]"
              type="text"
              :required="item.key in requiredValues"
              :state="inputState[item.key] ? inputState[item.key].state[currentLanguage] : null"
              class="form-control"
            />
          </b-form-group>
        </b-col>
      </b-row>
      <!-- Starting / finishing years are edited on first step only -->
      <b-row v-if="currentStep === 0">
        <b-col
          v-for="key in ['startYear', 'endYear']"
          :key="key"
          cols="6"
        >
          <small :for="'input-'+ kebabCase(key)">{{ formData.find(item => item.key === key).label }}</small>
          <b-form-group
            :id="'edit-' + kebabCase(key)"
            :invalid-feedback="invalidFeedback(key)"
          >
            <b-input
              :id="'input-' + kebabCase(key)"
              v-model="editedValues[key]"
              type="number"
              :min="1970"
              :max="new Date().getFullYear()"
              :description="formData.find(item => item.key === key).descriptions[currentLanguage]"
              :state="editedValues[key] ? inputState[key].state : null"
            />
          </b-form-group>
        </b-col>
      </b-row>
    </div>
    <div v-else>
      <b-card-group
        v-for="langKey in Object.keys(usedLanguages)"
        :key="'preview-in-' + langKey"
        deck
      >
        <EducationCard
          :education-data="editedValues"
          :language="langKey"
        />
      </b-card-group>
    </div>
    <b-btn
      id="save"
      class="pull-right"
      variant="success"
      :disabled="!canContinue"
      @click="handleNextButton()"
    >
      {{ currentStep === steps.length - 1 ? 'Save' : 'Next' }}
    </b-btn>
    <b-btn
      id="cancel"
      type="button"
      variant="light"
      class="pull-right"
      @click="handleBackButton()"
    >
      {{ currentStep === 0 ? 'Cancel' : 'Back' }}
    </b-btn>
  </div>
</template>

<script>
import { kebabCase, cloneDeep } from 'lodash'
import { EducationCard } from '@/components/Common'

export default {
  name: 'CvInfoEditEducationForm',
  components: { EducationCard },
  props: {
    initialValues: {
      type: Object,
      required: true
    }
  },
  data () {
    return {
      editedValues: cloneDeep(this.initialValues),
      requiredValues: ['school'],
      currentStep: 0,
      usedLanguages: {
        fi: { label: 'Suomeksi' },
        en: { label: 'in English' }
      },
      formData: [
        { key: 'school',
          label: 'Place of education',
          descriptions: {
            fi: 'Oppilaitoksen nimi, esim. Aalto-yliopisto',
            en: 'Name of the shool or institution, e.g. Aalto University'
          }
        },
        { key: 'degree',
          label: 'Degree',
          descriptions: {
            fi: 'Tutkinnon nimi, esim. Tietotekniikan diplomi-insinööri',
            en: 'e.g. M.Sc. (tech) in Computer Science'
          }
        },
        { key: 'major',
          label: 'Major',
          descriptions: {
            fi: 'Pääaineen/suuntautumisen nimi, esim. Ohjelmistotekniikka',
            en: 'Name of the major/study track, e.g Computer Science'
          }
        },
        { key: 'minor',
          label: 'Minor',
          descriptions: {
            fi: 'Sivuaineen nimi, esim. Geoinformatiikka',
            en: 'Name of the minor studied, e.g. Geoinformatics'
          }
        },
        { key: 'startYear', label: 'Starting year', descriptions: { fi: 'Aloitusvuosi', en: '' } },
        { key: 'endYear', label: 'Finishing year', descriptions: { fi: 'Lopetusvupsi', en: '' } }

      ],
      steps: [
        { languageKey: 'fi', useFormTemplate: true, stepTitle: 'Fill data in Finnish' },
        { languageKey: 'en', useFormTemplate: true, stepTitle: 'Fill data in English' },
        { useFormTemplate: false, stepTitle: 'Review and submit' }
      ]
    }
  },
  computed: {
    currentLanguage () {
      return this.steps[this.currentStep].languageKey
    },
    startYearState () {
      const rules = [
        {
          state: this.editedValues.endYear ? Number(this.editedValues.startYear) <= Number(this.editedValues.endYear) : true,
          feedback: 'Starting year cannot be after finishing year'
        },
        this.isValidYear(this.editedValues.startYear)
      ]
      return {
        state: rules.every(item => item.state),
        feedback: rules.filter(item => !item.state).map(item => item.feedback).join('\n')
      }
    },
    inputState () {
      return {
        school: {
          state: {
            fi: this.editedValues.fi.school.length > 0,
            en: this.editedValues.en.school.length > 0
          },
          feedback: 'Cannot be empty'
        },
        startYear: this.startYearState,
        endYear: this.isValidYear(this.editedValues.endYear)
      }
    },
    canContinue () {
      if (this.currentStep < this.steps.length) {
        return this.formIsValidByLanguage(this.steps[this.currentStep].languageKey)
      } else {
        return Object.keys(this.usedLanguages).every(key => this.formIsValidByLanguage(key))
      }
    }
  },
  watch: {
    initialValues: (val) => {
      this.editedValues = Object.assign({}, val)
    }
  },
  methods: {
    kebabCase (text) {
      return kebabCase(text)
    },
    isValidYear (year) {
      return {
        state: year ? Number(year) >= 1970 && Number(year) <= new Date().getFullYear() : true,
        feedback: `Invalid year, should be between 1970-${new Date().getFullYear()}`
      }
    },
    invalidFeedback (key, langKey = null) {
      if (this.inputState[key]) {
        const state = this.inputState[key].state
        // Handle both cases: state.[lang] == true and state == true
        if (state[langKey] || state === true) {
          return ''
        } else {
          return this.inputState[key].feedback
        }
      }
      return ''
    },
    formIsValidByLanguage (langKey) {
      const statesForLanguage = Object.keys(this.inputState).map(key => {
        const item = this.inputState[key]
        if (item.state.hasOwnProperty(langKey)) {
          return item.state[langKey]
        } else {
          return item.state
        }
      })
      return statesForLanguage.every(state => state)
    },
    handleNextButton () {
      if (this.currentStep < this.steps.length - 1) {
        ++this.currentStep
      } else {
        this.$emit('submit', this.editedValues)
      }
    },
    handleBackButton () {
      if (this.currentStep > 0) {
        --this.currentStep
      } else {
        this.editedValues = Object.assign({}, this.initialValues)
        this.$emit('cancel')
      }
    }
  }
}
</script>

<style scoped >
.step-title {
  font-weight: 500;
  font-size: 15px;
  text-transform: uppercase;
}
.step-title.current-step {
  font-weight: bolder;
}
</style>
