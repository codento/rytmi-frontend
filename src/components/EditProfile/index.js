import EditLinks from './EditLinks.vue'
import EditSkills from './EditSkills.vue'
import ProfileForm from './ProfileForm.vue'
import EditProjects from './EditProjects.vue'
import WorkHistoryEntryList from './WorkHistoryEntryList.vue'
import WorkHistoryEntryEditList from './WorkHistoryEntryEditList'
import CvInfo from './CvInfo.vue'
import CvInfoEditIntroduction from './CvInfoEditIntroduction.vue'
import CvInfoEditEducation from './CvInfoEditEducation.vue'
import CvInfoEditEducationForm from './CvInfoEditEducationForm.vue'

export {
  EditLinks,
  EditSkills,
  ProfileForm,
  EditProjects,
  WorkHistoryEntryList,
  CvInfo,
  CvInfoEditIntroduction,
  CvInfoEditEducation,
  CvInfoEditEducationForm,
  WorkHistoryEntryEditList
}
