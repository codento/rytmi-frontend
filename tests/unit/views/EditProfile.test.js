import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import BootstrapVue from 'bootstrap-vue'
import EditProfile from '@/views/EditProfile'
import {
  EditLinks,
  EditSkills,
  EditProjects,
  ProfileForm
} from '@/components/EditProfile'
import { merge } from 'lodash'

const localVue = createLocalVue()
localVue.use(BootstrapVue)
localVue.use(Vuex)

const mockProfile = { id: 5 }

function createStore (overrideConfig) {
  const getters = {
    profileById: () => () => mockProfile
  }
  const mergedStoreConfig = merge({ getters }, overrideConfig)
  return new Vuex.Store(mergedStoreConfig)
}

function createWrapper (overrideOptions) {
  const mountOptions = {
    localVue,
    store: createStore()
  }
  const mergedOptions = merge(mountOptions, overrideOptions)
  return shallowMount(EditProfile, mergedOptions)
}

describe('EditProfile.vue', () => {
  it('should not display any subcomponents when profile cannot be found', () => {
    const propsData = {
      profileId: 'unknownprofile'
    }
    const store = createStore({
      getters: {
        profileById: () => () => null
      }
    })
    const wrapper = createWrapper({ propsData, store })
    expect(wrapper.find(ProfileForm).exists()).toBeFalsy()
    expect(wrapper.find(EditLinks).exists()).toBeFalsy()
    expect(wrapper.find(EditSkills).exists()).toBeFalsy()
    expect(wrapper.find(EditProjects).exists()).toBeFalsy()
  })

  it('Should show correct components when profileId is not null', () => {
    const propsData = {
      profileId: '5'
    }
    const wrapper = createWrapper({ propsData })
    const profileFormWrapper = wrapper.find(ProfileForm)
    const editSkillsWrapper = wrapper.find(EditSkills)
    const editLinksWrapper = wrapper.find(EditLinks)
    const EditProjectsWrapper = wrapper.find(EditProjects)
    expect(profileFormWrapper.isVisible()).toBeTruthy()
    expect(profileFormWrapper.props().profile).toBe(mockProfile)
    expect(editSkillsWrapper.isVisible()).toBeTruthy()
    expect(editSkillsWrapper.props().profileId).toBe(mockProfile.id)
    expect(editLinksWrapper.isVisible()).toBeTruthy()
    expect(editLinksWrapper.props().profile).toBe(mockProfile)
    expect(EditProjectsWrapper.isVisible()).toBeTruthy()
    expect(EditProjectsWrapper.props().profileId).toBe(mockProfile.id)
    expect(wrapper.element).toMatchSnapshot()
  })
})
