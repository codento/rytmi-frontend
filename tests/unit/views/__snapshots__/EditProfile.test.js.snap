// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfile.vue Should show correct components when profileId is not null 1`] = `
<div
  class="animated fadeIn profile-editor col-sm-12 col-md-10"
>
  <h1>
    Edit Profile
  </h1>
   
  <hr />
   
  <h1>
    Basic Details
  </h1>
   
  <hr />
   
  <profileform-stub
    profile="[object Object]"
  />
   
  <cvinfo-stub
    profile="[object Object]"
  />
   
  <workhistoryentrylist-stub
    profileid="5"
  />
   
  <editlinks-stub
    profile="[object Object]"
  />
   
  <editskills-stub
    profileid="5"
  />
   
  <editprojects-stub
    profileid="5"
  />
</div>
`;
